### ISSUE TITLE
銀行のテレマーケティング

### ISSUE THREAD

**Issue by aaaa383** [aaaa383/AIquest#2]
#モジュールのインポート
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
%matplotlib inline
from sklearn.tree import DecisionTreeClassifier as DT
from sklearn.model_selection import cross_validate
from sklearn.model_selection import GridSearchCV

from sklearn.metrics import roc_auc_score

#データの読み込み
train = pd.read_csv("../input/train.csv")
test = pd.read_csv("../input/test.csv")
submit = pd.read_csv("../input/sample_submission.csv", header=None)

#treinデータの確認
train.head()
#testデータの確認
test.head()

#trainデータの基礎統計量の集計
train.describe()
#testデータの基本統計量の集計
test.describe()

print( train.describe(include=['O']) )
print( test.describe(include=['O']) )

#trainの行、列の数の確認
train.shape
#testの行、列の数の確認
test.shape

#カラムの確認
print(train.info())
print(test.info())

# カラムpoutcomeの要素の種類と出現数の表示
print( train['poutcome'].value_counts() )

# カラムyの要素の種類と出現数の表示
print( train['y'].value_counts() )

cross = pd.crosstab(train['poutcome'], train['y'], margins=True)

# クロス集計結果の表示
print( cross )

# 申込率の計算
rate = cross[1] / cross['All']

# クロス集計結果に申込率cvrを追加
cross['cvr'] = rate

# 申込率を追加したクロス集計表の表示
print( cross )

# クロス集計表からインデックス'success', 'failure'、カラム'cvr'のみを取り出して表示
print( cross.loc[['success', 'failure'], 'cvr'])

#相関
corr_matrix = train.corr()
print(corr_matrix)

#ヒートマップの作成
import seaborn as sns

# heatmapを作成
sns.heatmap(corr_matrix, cmap="Reds")

# グラフにタイトルを追加
plt.title('Correlation')

# グラフを表示
plt.show()

#durationの抜き出し
duration_0 = train[train['y']==0]['duration']
duration_1 = train[train['y']==1]['duration']

# ヒストグラムの作成
sns.distplot(duration_0, label='y=0')
sns.distplot(duration_1, label='y=1')

# グラフにタイトルを追加
plt.title('duration histgram')

# グラフのx軸に名前を追加
plt.xlabel('duration')

# グラフのy軸に名前を追加
plt.ylabel('frequency')

# x軸の表示範囲の指定
plt.xlim(0, 2000)

# グラフに凡例を追加
plt.legend()

# グラフを表示
plt.show()

trainX = train.drop('y', axis=1)
trainX.head()

y = train["y"]
y.head()
testX = test.copy()

#ダミー変数化
trainX = pd.get_dummies(trainX)
testX = pd.get_dummies(testX)

#モデルの箱を作成
tree = DT(random_state=0)

#パラメータ
parameters = {"max_depth": [16,17,18,19,20,21],"min_samples_leaf": [60,65,70,75,80],"min_samples_split":[2,3,4,5]}

#グリッドサーチ
gcv = GridSearchCV(tree,parameters,cv=5,scoring="roc_auc",n_jobs=-1)
gcv.fit(trainX,y)

#最適なパラメータ
gcv.best_params_

# 最適なパラメータで学習したモデルの取得
best_model = gcv.best_estimator_

# 評価用データの予測
pred_y = best_model.predict_proba(testX)[:,1]

# 決定木描画ライブラリのインポート
from sklearn.tree import export_graphviz

export_graphviz(best_model, out_file="tree.dot", feature_names=train_X.columns, class_names=["0","1"], filled=True, rounded=True)

# 決定木グラフの表示
from matplotlib import pyplot as plt
from PIL import Image
import pydotplus
import io
g = pydotplus.graph_from_dot_file(path="tree.dot")
gg = g.create_png()
img = io.BytesIO(gg)
img2 = Image.open(img)
plt.figure(figsize=(img2.width/100, img2.height/100), dpi=100)
plt.imshow(img2)
plt.axis("off")
plt.show()

#提出用ファイルの作成
submit[1] = pred_y
submit.to_csv("submit.csv", index=False, header=False)

**Comment by aaaa383** [comment]
signate社のコンペのモデルです。
評価指標はAUCで結果は　0.8942834631525661 です。
